curvefs/src/metaserver/copyset/concurrent_apply_queue.cpp

```diff
@@ -0,0 +1,200 @@
+    InitThreadPool(ThreadPoolType::WRITE, wconcurrentsize_, wqueuedepth_);
+
+    if (!cond_.WaitFor(5000)) {
+        LOG(ERROR) << "init concurrent module's threads fail";
+        start_ = false;
+    }
+
+    LOG(INFO) << "Init concurrent module's threads success";
+    return start_;
+}
+
+bool ConcurrentApplyModule::checkOptAndInit(
+    const ConcurrentApplyOption &opt) {
+    if (opt.rconcurrentsize <= 0 || opt.wconcurrentsize <= 0 ||
+        opt.rqueuedepth <= 0 || opt.wqueuedepth <= 0) {
+        LOG(INFO) << "init concurrent module fail, params must >=0"
+            << ", rconcurrentsize=" << opt.rconcurrentsize
+            << ", wconcurrentsize=" << opt.wconcurrentsize
+            << ", rqueuedepth=" << opt.rqueuedepth
+            << ", wconcurrentsize=" << opt.wqueuedepth;
+        return false;
+    }
+
+    wconcurrentsize_ = opt.wconcurrentsize;
+    wqueuedepth_ = opt.wqueuedepth;
+    rconcurrentsize_ = opt.rconcurrentsize;
+    rqueuedepth_ = opt.rqueuedepth;
+
+    return true;
+}
+
+
+void ConcurrentApplyModule::InitThreadPool(
+    ThreadPoolType type, int concurrent, int depth) {
+    for (int i = 0; i < concurrent; i++) {
+        auto asyncth = new (std::nothrow) TaskThread(depth);
+        CHECK(asyncth != nullptr) << "allocate failed!";
+
+        switch (type) {
+        case ThreadPoolType::READ:
+            rapplyMap_.insert(std::make_pair(i, asyncth));
+            break;
+
+        case ThreadPoolType::WRITE:
+            wapplyMap_.insert(std::make_pair(i, asyncth));
+            break;
+        }
+    }
+
+    for (int i = 0; i < concurrent; i++) {
+        switch (type) {
+        case ThreadPoolType::READ:
+            rapplyMap_[i]->th =
+                    std::thread(&ConcurrentApplyModule::Run, this, type, i);
+            break;
+
+        case ThreadPoolType::WRITE:
+            wapplyMap_[i]->th =
+                    std::thread(&ConcurrentApplyModule::Run, this, type, i);
+            break;
+        }
+    }
+}
+
+void ConcurrentApplyModule::Run(ThreadPoolType type, int index) {
+    cond_.Signal();
+    while (start_) {
+        switch (type) {
+        case ThreadPoolType::READ:
+            rapplyMap_[index]->tq.Pop()();
+            break;
+
+        case ThreadPoolType::WRITE:
+            wapplyMap_[index]->tq.Pop()();
+            break;
+        }
+    }
+}
+
+void ConcurrentApplyModule::Stop() {
+    if (!start_.exchange(false)) {
+        return;
+    }
+
+    LOG(INFO) << "stop ConcurrentApplyModule...";
+    auto wakeup = []() {};
+    for (auto iter : rapplyMap_) {
+        iter.second->tq.Push(wakeup);
+        iter.second->th.join();
+        delete iter.second;
+    }
+    rapplyMap_.clear();
+
+    for (auto iter : wapplyMap_) {
+        iter.second->tq.Push(wakeup);
+        iter.second->th.join();
+        delete iter.second;
+    }
+    wapplyMap_.clear();
+
+    LOG(INFO) << "stop ConcurrentApplyModule ok.";
+}
+
+void ConcurrentApplyModule::Flush() {
```

Comment:
you should flush the queue when current copyset becomes a leader fo this group.

Reply:
done